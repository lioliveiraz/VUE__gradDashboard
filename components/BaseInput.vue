<template>
  <div class="g-base-input-wrapper">
    <label data-testId="label">{{ attributeObj.name.toUpperCase() }}</label>
    <input
      data-testId="input"
      :class="
        checkObjectEmpty
          ? 'g-base-input border-green-400	'
          : 'g-base-input  border-red-500'
      "
      :type="attributeObj.type"
      :name="attributeObj.name"
      :placeholder="attributeObj.placeholder"
      autofocus
      v-model="currentValue"
      :required="attributeObj.required"
    />
    <p class="text-red-700 font-bold">{{ errorMessage }}</p>
  </div>
</template>

<script>
import { isObjectEmpty } from "@/helpers/service";
import { userValidation } from "@/helpers/validation";
import { courseValidation } from "../helpers/validation";
import _ from "lodash";

export default {
  props: { attributeObj: Object },
  data() {
    return {
      currentValue: "",
      userError: { initialError: "error" },
      courseError: { initialError: "error" },
      errorMessage: "",
    };
  },
  computed: {
    checkObjectEmpty() {
      if (!this.currentValue) return false;
      else {
        return !isObjectEmpty(this.userError) ||
          !isObjectEmpty(this.courseError)
          ? false
          : true;
      }
    },
  },
  watch: {
    currentValue: function () {
      let inputObj = {};
      inputObj = { [this.attributeObj.name]: this.currentValue };
      this.userError = userValidation(inputObj);
      this.courseError = courseValidation(inputObj);
      this.$emit("getUserInput", this.currentValue, this.attributeObj.name);
      this.debouncedGetError();
    },
  },
  created: function () {
    this.debouncedGetError = _.debounce(this.getErrorMessage, 800);
  },
  methods: {
    getErrorMessage() {
       this.errorMessage =
        this.userError[this.attributeObj.name] ||
        this.courseError[this.attributeObj.name];
    },
  },
};
</script>

<style>
</style>