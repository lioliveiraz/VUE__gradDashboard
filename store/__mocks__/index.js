import Vue from 'vue';
import Vuex from 'vuex';

Vue.use(Vuex);

export const getters = {
    isLoggedIn: jest.fn().mockReturnValue(true),
    getToken: jest.fn().mockReturnValue({ token: "token" }),
    getName: jest.fn().mockReturnValue(),
    getCourses: jest.fn().mockReturnValue([{ name: "course" }]),
    getAssessments: jest.fn().mockReturnValue([{ name: "assessments" }]),
    getScores: jest.fn().mockReturnValue([{ name: "scores" }])

};

export const mutations = {
    SET_TOKEN: jest.fn(),
    SET_ID: jest.fn(),
    SET_NAME: jest.fn(),
    SET_COURSES: jest.fn(),
    SET_ASSESSMENTS: jest.fn(),
    SET_SCORES: jest.fn(),
    ADD_COURSE: jest.fn(),
    ADD_ASSESSMENT: jest.fn(),
    ADD_SCORE: jest.fn()
};

export const actions = {
    login: jest.fn(),
    logout: jest.fn(),
    fetchCourses: jest.fn(),
    fetchScores: jest.fn(),
    handleAddCourse: jest.fn(),
    handleAddScore: jest.fn()


};

export const state = {
    token: "token",
    user_id: 1,
    user_name: "name",
    courses: [{ name: "course" }],
    assessments: [{ name: "assessments" }],
    scores: [{ name: "scores" }]
};

export function __createMocks(custom = { getters: {}, mutations: {}, actions: {}, state: {} }) {
    const mockGetters = Object.assign({}, getters, custom.getters);
    const mockMutations = Object.assign({}, mutations, custom.mutations);
    const mockActions = Object.assign({}, actions, custom.actions);
    const mockState = Object.assign({}, state, custom.state);

    return {
        getters: mockGetters,
        mutations: mockMutations,
        actions: mockActions,
        state: mockState,
        store: new Vuex.Store({
            modules: {
                auth: {
                    namespaced: true,
                    getters: mockGetters,
                    mutations: mockMutations,
                    actions: mockActions,
                    state: mockState,
                },
                courses: {
                    namespaced: true,
                    getters: mockGetters,
                    mutations: mockMutations,
                    actions: mockActions,
                    state: mockState,
                }
            }

        }),
    };
}

export const store = __createMocks().store;