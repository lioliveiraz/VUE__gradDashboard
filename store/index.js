import Vuex from 'vuex';
import Vue from 'vue';
import auth from './modules/auth';

Vue.use(Vuex);


const store = () => {
    return new Vuex.Store({
        modules: { auth }
    });
};

export default store;