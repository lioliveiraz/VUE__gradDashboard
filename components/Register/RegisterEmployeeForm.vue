<template>
  <form @submit="handleSubmit" :key="key" class="g-form-wrapper">
    <div class="g-form-wrapper--inner">
      <BaseInput
        @getUserInput="getUserInput"
        :attributeObj="{
          type: this.TEXT_INPUT,
          name: this.EMPID_INPUT,
          placeholder: 'employer ID',
           label:'emp id',
          required: true,
        }"
      />

      <BaseInput
        @getUserInput="getUserInput"
        :attributeObj="{
          type: this.PASSWORD_INPUT,
          name: this.PASSWORD_INPUT,

          required: true,
        }"
      />

      <BaseInput
        @getUserInput="getUserInput"
        :attributeObj="{
          type: this.TEXT_INPUT,
          name: this.EMP_NAME_INPUT,
          placeholder: 'employer name',
        
        }"
      />

      <input
        :type="this.BUTTON_SUBMIT"
        :class="isFormValid ? 'g-base-btn-green' : 'g-base-btn-green g-disabled'"
        :disabled="!isFormValid"
      />
    </div>
  </form>
</template>

<script>
import { mapGetters } from "vuex";
import { registerEmployee } from "../../api/requests/post";
import { userValidation } from "../../helpers/validation";
import BaseInput from "../BaseInput";
import {formUserMixin} from '@/mixin/formUserMixin'

export default {
  components: { BaseInput },
  mixins:[formUserMixin],
  data() {
    return {
      formData: {},
      key: 0,
      isFormValid: false,
    };
  },
  computed: {
    ...mapGetters("auth", ["getToken"]),
  },
  methods: {
    async handleSubmit(e) {
      e.preventDefault();
      const errors = userValidation(this.formData);
      if (!this.isFormValid) return null;

      try {
        const res = await registerEmployee(this.formData, this.getToken);
        this.errors = {};

        this.$toast(res.data.message, { type: this.TOAST_SUCCESS });
        this.key = this.key + 1;
      } catch (err) {
        this.$toast(err.response.data.message, { type: this.TOAST_ERROR });
      }
    },
  },
};
</script>

<style scoped>

</style>