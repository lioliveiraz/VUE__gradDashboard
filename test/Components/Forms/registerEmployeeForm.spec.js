import Vuex from 'vuex';
import RegisterEmployeeForm from '../../../components/Register/RegisterEmployeeForm.vue';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import { __createMocks, store } from '../../../store/__mocks__';
import BaseInput from '../../../components/BaseInput.vue';


jest.mock('../../../store/__mocks__');

const localVue = createLocalVue();

localVue.use(Vuex);

describe('<RegisterEmployeeForm/>', () => {
    let wrapper, mockedData, errors;
    mockedData = {

    };
    errors = {};
    beforeEach(async () => {
        wrapper = await shallowMount(RegisterEmployeeForm, {
            store: store,
            localVue
        });
    });
    it('should render correctly', () => {
        const idArr = [".g-form-wrapper", ".g-form-wrapper--inner", '.g-base-btn-submit'];
        idArr.forEach((id) => expect(wrapper.find(id)).toBeTruthy());
        expect(wrapper).toMatchSnapshot();

    });
    it('child components render correctly', () => {
        const baseInput = wrapper.findComponent(BaseInput);
        const inputs = wrapper.findAllComponents(BaseInput);

        expect(baseInput.exists()).toBeTruthy();
        expect(inputs).toHaveLength(3);

    });
    it('data initialize correctly', () => {

        expect(RegisterEmployeeForm.data().employeeObj).toEqual(mockedData);
        expect(RegisterEmployeeForm.data().errors).toEqual({});
        expect(RegisterEmployeeForm.data().key).toEqual(0);

    });

    it("should return an object with errors", async () => {
        const form = wrapper.find('form');

        mockedData = {
            name: "!",
            empId: "!"
        };

        await wrapper.setData({
            employeeObj: mockedData
        });

        await form.trigger("submit");
        errors = {
            name: expect.any(String),
            empId: expect.any(String)
        };
        expect(wrapper.vm.errors).toEqual(errors);
    });

    it("should NOT return an object with errors", async () => {
        const form = wrapper.find('form');

        mockedData = {
            name: "name",
            empId: "123456"
        };

        await wrapper.setData({
            employeeObj: mockedData
        });

        await form.trigger("submit");
        errors = {
        };
        expect(wrapper.vm.errors).toEqual(errors);

    });

    it("should change the date once the method is called", async () => {
        await wrapper.vm.getUserInput("value", "name");
        let object = {
            name: "value"
        };
        expect(wrapper.vm.employeeObj).toEqual(expect.objectContaining(object));

    });


});