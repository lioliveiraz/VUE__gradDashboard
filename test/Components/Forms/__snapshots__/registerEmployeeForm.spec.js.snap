// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RegisterEmployeeForm/> should render correctly 1`] = `
VueWrapper {
  "_emitted": Object {},
  "_emittedByOrder": Array [],
  "isFunctionalComponent": undefined,
}
`;
